lambda/matching.ml

```diff
@@ -2686,41 +2709,61 @@ let combine_constant value_kind loc arg cst partial ctx def
 let split_cases tag_lambda_list =
   let rec split_rec = function
     | [] -> ([], [])
-    | (cstr_tag, act) :: rem -> (
+    | ({cstr_tag; cstr_repr; cstr_constant}, act) :: rem -> (
```

Comment:
This code previously took a list of tags (corresponding to variant constructors) and actions, and sorted them by whether the variant constructor being matched on has arguments or not.  Tags used to have enough information to make this determination, but we've simplified the tag type and moved some of the information it held into a new `variant_representation` type, which is more uniform with the treatment of records.

Since tags no longer have enough information to determine whether the variant constructors are constant or not, I changed this to take a `variant_representation` as well (and `cstr_constant` in the `boxed` case, because it seems wasteful to constantly check whether all the fields are void in the boxed case).  `split_variant_cases` has been added for the same reason: `divide_variant` used to use a tag to communicate constant/non-constant information to `split_cases`, but it seemed cleaner to make a special-purpose `split_variant_cases` than to have `split_variant_cases` make a fake `variant_representation` so it could work with `split_cases`.

The pattern match compiler is tricky, so it seems sensible to get another reviewer or go through it carefully if this isn't clarifying.